core/src/main/java/io/grpc/internal/RetryScheduler.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+/**
+ * Schedules a retry operation according to a {@link BackoffPolicy}. The retry is run withing a
+ * {@link SynchronizationContext}. At most one retry is scheduled at a time.
+ */
+final class RetryScheduler {
+  private final Runnable retryOperation;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final SynchronizationContext syncContext;
+  private final BackoffPolicy.Provider policyProvider;
+
+  private BackoffPolicy policy;
+  private ScheduledHandle scheduledHandle;
+
+  private static final Logger logger = Logger.getLogger(RetryScheduler.class.getName());
+
+  RetryScheduler(Runnable retryOperation, ScheduledExecutorService scheduledExecutorService,
+      SynchronizationContext syncContext, BackoffPolicy.Provider policyProvider) {
+    this.retryOperation = retryOperation;
+    this.scheduledExecutorService = scheduledExecutorService;
+    this.syncContext = syncContext;
+    this.policyProvider = policyProvider;
+  }
+
+  /**
+   * Schedules a future retry operation. Only allows one retry to be scheduled at any given time.
+   *
+   * @return The delay in nanos before the operation fires or -1 if it was not scheduled.
+   */
+  long schedule() {
+    if (policy == null) {
+      policy = policyProvider.get();
+    }
+    // If a retry is already scheduled, take no further action.
+    if (scheduledHandle != null && scheduledHandle.isPending()) {
+      return -1;
+    }
+    long delayNanos = policy.nextBackoffNanos();
```

Comment:
I guess I could, but if the task can't be scheduled for some reason we would get a RejectedExecutionException which is pretty good at conveying what wen't wrong. Since the existing code to schedule retries didn't make this upfront check either, I'm thinking it's ok the way it is. Let me know if you have a stronger opinion on it.